### PR TITLE
MAINT: stats.fisher_exact: improve docs and fix bugs

### LINKED ISSUE #4130: BUG: stats: fisher_exact returns incorrect p-value
For some tables, `fisher_exact` incorrectly returns a p-value of 1.  For example,

```
In [1]: from scipy.stats import fisher_exact

In [2]: m = np.array([[345, 455], [260, 345]])

In [3]: fisher_exact(m)
Out[3]: (1.0061284868977176, 1.0)
```

The correct p-value for the array `m` is 0.9567.   Here, for example, is the result in R:

```
> m <- array(c(345, 455, 260, 345), c(2,2))
> m
     [,1] [,2]
[1,]  345  260
[2,]  455  345
> fisher.test(m)

    Fisher's Exact Test for Count Data

data:  m
p-value = 0.9567
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
 0.8083048 1.2526776
sample estimates:
odds ratio 
  1.006125 
```

The culprit is this code, from the `fisher_exact` function in `scipy/stats/stats.py`:

```
    elif alternative == 'two-sided':
        mode = int(float((n + 1) * (n1 + 1)) / (n1 + n2 + 2))
        pexact = hypergeom.pmf(c[0,0], n1 + n2, n1, n)
        pmode = hypergeom.pmf(mode, n1 + n2, n1, n)

        epsilon = 1 - 1e-4
        if np.abs(pexact - pmode) / np.maximum(pexact, pmode) <= 1 - epsilon:
            return oddsratio, 1.
```

In effect, it says the p-value is 1 if `np.abs(pexact - pmode) / np.maximum(pexact, pmode) <= 1e-4`.  It isn't clear why the value 1e-4 was used here.

### LINKED ISSUE #9231: infinite loop in stats.fisher_exact
I'm using scipy version 1.0.0.

```
import scipy as sp
import scipy.stats
x = [[5829225, 5692693], [5760959, 5760959]]
sp.stats.fisher_exact(x)

```
For the values above scipy does not return anything but waits. What can be the reason for that? How can I fix it?

However in R it returns a p-value almost immediately.

```
a = matrix(c(5829225,5692693,5760959,5760959), nrow=2)
fisher.test(a)
```

Thanks in advance, 
Burcak

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-4130
Closes gh-9231

#### What does this implement/fix?
gh-4130 was originally about inaccurate p-values due to a large fudge factor of 1e-4 used in the implementation. IIUC, the fudge factor is a relative tolerance used to compare numerically distinct but theoeretically equal probability masses. The `hypergeom` distribution is from Boost now, so it is more accurate that what we had before. Perhaps 1e-14 is too tight, but in this case, I think we're looking at small inaccuracies in the reported p-value.

Most of the discussion in that issue is actually about confusion surrounding the definition of two-sided p-value, so I added some additional information to the documentation, including a more intutitive (but also accurate) explanation of the null hypothesis.

gh-9231 exposed a bug that made `fisher_exact` very slow in some cases. This PR fixes it by replacing the binary search in `fisher_exact` with the one written for `binomtest`.

### CODE REVIEW

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4477]: I would prefer to make `epsilon` be `1e-14` and use `1 - epsilon` elsewhere, but I didn't want to confuse this PR with a refactoring.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4476]: Feel free to suggest a different value, and I'll make the change. When I thought of `1e-14`, I was thinking this was an absolute tolerance, but it's relative, and it might be too tight for very small probability masses. [`binomtest`](https://github.com/scipy/scipy/blob/b5d8bab88af61d61de09641243848df63380a67f/scipy/stats/_binomtest.py#L301) uses something like `1e-7` and [this license-compatible implementation](https://github.com/brentp/fishers_exact_test/blob/master/src/cfisher.pyx) uses `1e-6`, but they are not using the tolerance in exactly the same way.

**Review by tupui — APPROVED** [review]
Thanks @mdhaber. It reads good to me. I think in general we should think about how to present in a more consistent way the limitation and interpretation of our tests.

I will wait for some green CI before merging.

File: scipy/stats/_stats_py.py
Comment (tupui) [reply scipy/stats/_stats_py.py:4476]: Since you said you will update this part later, fine with me like this.

File: scipy/stats/tests/test_stats.py
Comment (tupui) [inline scipy/stats/tests/test_stats.py:547]:
```suggestion
        x = [[6, 37], [108, 200]]
```

**Comment by mdhaber** [comment]
> I think in general we should think about how to present in a more consistent way the limitation and interpretation of our tests.

It got better when we added descriptions of alternatives to all the hypothesis tests (except I see we missed ones for contingency tables). But yes, it would be better to be more consistent to address the null hypothesis, statistic, and p-value in the introduction of each test, and I'm typically adding these as I work on functions. I started putting together a spreadsheet earlier:

https://docs.google.com/spreadsheets/d/1YtO7YrQ7TU4vjuVDtY-lL3UK7Za-EQCiTnYoSBhqxTM/edit?usp=sharing

but I decided not to post a tracking issue. Feel free to do so if you're interested, though.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4241]: Hold off on merge, actually. I need to figure out how to work in something about the marginals of the table.

**Comment by mdhaber** [comment]
Thanks again @tupui. I added the condition that the marginals of the tables sampled must equal those of the observed table. I chose to state this condition along with the null hypothesis, whereas some might argue for it to be specified only when defining the the p-value. Either way, I think it is an improvement for the relationship between the null hypothesis, the alternatives, the statistic, and the p-values to be stated like this rather than solely in the notes in terms of the hypergeometric distribution. We can sit on it, and refinements could be made in the followup PR that addresses gh-9231.

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4468]:
```suggestion
    return oddsratio, pvalue
```

File: scipy/stats/_stats_py.py
Comment (mdhaber) [inline scipy/stats/_stats_py.py:4242]:
```suggestion
    from these populations under a condition: the marginals of the
```

**Review by tupui — APPROVED** [review]
Update is looking good. I think it's fine for readers with a stats background but it's hard to read otherwise. If we add more things like these, we need to write some central tutorial about all the terminology we use here.

Good thing you did a table. Before going through I think we should discuss in detail how we want everything to look in the end, otherwise we are exposing ourselves to a lot of [re]work. One thing we can do is to work on a single test to sort everything out. Let's discuss this later today.

**Pull request merged**
